### PR TITLE
Essentials.FileSystem refactored to interface and singleton

### PR DESCRIPTION
**Pull request by moljac**
# WORK IN PROGRESS

### Description of Change ###

Implements #

### Additions made ###

* Adds 

### PR Checklist ###

- [ ] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)
- [ ] Targets a single property for a single control (or intertwined few properties)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the WinUI, Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might affect accessibility?
- [ ] Does this PR introduce a new control? (If yes, add an example using SemanticProperties to the SemanticsPage)
- [ ] APIs that modify focusability?
- [ ] APIs that modify any text property on a control?
- [ ] Does this PR modify view nesting or view arrangement in anyway?
- [ ] Is there the smallest possibility that your PR will change accessibility? 
- [ ] I'm not sure, please help me

If any of the above checkboxes apply to your PR, then the PR will need to provide testing to demonstrate that accessibility still works.

### CODE REVIEW

**Review by jsuarezruiz — COMMENTED** [review]
There are some build errors here:
```
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(83,34): error CS0506: 'BookmarkDataFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(172,34): error CS0506: 'UIDocumentFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(194,34): error CS0506: 'UIImageFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(215,34): error CS0506: 'PHAssetFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(145,4): error CS0103: The name 'Init' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(175,19): error CS1501: No overload for method 'GetContentType' takes 1 arguments [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(208,7): error CS0103: The name 'Current' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(83,34): error CS0506: 'BookmarkDataFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(172,34): error CS0506: 'UIDocumentFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(194,34): error CS0506: 'UIImageFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.ios.cs(215,34): error CS0506: 'PHAssetFileResult.OpenReadAsync()': cannot override inherited member 'FileBase.OpenReadAsync()' because it is not marked virtual, abstract, or override [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(145,4): error CS0103: The name 'Init' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(175,19): error CS1501: No overload for method 'GetContentType' takes 1 arguments [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(208,7): error CS0103: The name 'Current' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(145,4): error CS0103: The name 'Init' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(175,19): error CS1501: No overload for method 'GetContentType' takes 1 arguments [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.shared.cs(208,7): error CS0103: The name 'Current' does not exist in the current context [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.android.cs(11,50): error CS0736: 'FileSystemImplementation' does not implement instance interface member 'IFileSystem.CacheDirectory'. 'FileSystemImplementation.CacheDirectory' cannot implement the interface member because it is static. [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
/Users/builder/azdo/_work/2/s/src/Essentials/src/FileSystem/FileSystem.android.cs(11,50): error CS0736: 'FileSystemImplementation' does not implement instance interface member 'IFileSystem.AppDataDirectory'. 'FileSystemImplementation.AppDataDirectory' cannot implement the interface member because it is static. [/Users/builder/azdo/_work/2/s/src/Essentials/src/Essentials.csproj]
```

Ops, I have seen later that it is still in progress.

**Comment by moljac** [comment]
@jsuarezruiz 

This one was a bit thougher than the others. I pushed all working refactorings as PR and added H1 WORK IN PROGRESS (I have no sufficient rights to mark PR as draft, or label it or assign).

This was done merely so we would not double our work (syncing efforts).

File: src/Essentials/src/FileSystem/FileSystem.shared.cs [hunk 1]
```diff
@@ -87,6 +128,20 @@ internal static string Clean(string extension, bool trimLeadingPeriod = false)
 				return extension;
 			}
 		}
+
+#nullable enable
```
Comment (mattleibow) [inline src/Essentials/src/FileSystem/FileSystem.shared.cs:132]: The nullable scope should also cover Current, so ight as well use a single block.

File: src/Essentials/src/FileSystem/FileSystem.shared.cs [hunk 2]
```diff
@@ -1,26 +1,67 @@
...
+			=> (Current as IPlatformFileSystem)?.GetTemporaryFile(root, fileName);
+
+		internal static string EnsurePhysicalPath(Android.Net.Uri uri, bool requireExtendedAccess = true)
+			=> (Current as IPlatformFileSystem)?.EnsurePhysicalPath(uri, requireExtendedAccess);
+#endif
+
+#if IOS || MACCATALYST
+		internal static Task<FileResult[]> EnsurePhysicalFileResultsAsync(params Foundation.NSUrl[] urls)
+			=> (Current as IPlatformFileSystem)?.EnsurePhysicalFileResultsAsync(urls);
+#endif
```
Comment (mattleibow) [inline src/Essentials/src/FileSystem/FileSystem.shared.cs:63]: I wonder if these should throw? Especially the async... a null Task is very unexpected.

File: src/Essentials/src/FileSystem/FileSystem.netstandard.cs [hunk 3]
```diff
@@ -1,34 +1,41 @@
...
+namespace Microsoft.Maui.Essentials
+{
 	/// <include file="../../docs/Microsoft.Maui.Essentials/FileBase.xml" path="Type[@FullName='Microsoft.Maui.Essentials.FileBase']/Docs" />
 	public partial class FileBase
 	{
 		static string PlatformGetContentType(string extension) =>
 			throw ExceptionUtils.NotSupportedOrImplementedException;
 
-		internal void PlatformInit(FileBase file) =>
+		internal void Init(FileBase file) =>
```
Comment (mattleibow) [inline src/Essentials/src/FileSystem/FileSystem.netstandard.cs:31]: What happened here? And why the blank func below?

**Review by mattleibow — COMMENTED** [review]
Just going to mention it for completeness, but I think this is different, should the return `FileResult` be `IFileResult` (like the screenshots api #4699)

This might be different as `FileResult` might very well be a representation of a file on disk. However, what happens if some unit testing system or platform wants/needs to represent a file system as a virtual one? Like say a single blob with offsets?

**Review by Redth — APPROVED** [review]

**Pull request merged**